src/cargo/core/source_id.rs

```diff
@@ -666,15 +666,12 @@ impl fmt::Display for SourceId {
 }
 
 /// The hash of SourceId is used in the name of some Cargo folders, so shouldn't
-/// vary. `as_str` gives the serialisation of a url (which has a spec) and so
+/// vary. [`CanonicalUrl`] uses `as_str`, the serialisation of a url (which has a spec), and so
 /// insulates against possible changes in how the url crate does hashing.
 impl Hash for SourceId {
     fn hash<S: hash::Hasher>(&self, into: &mut S) {
         self.inner.kind.hash(into);
-        match self.inner.kind {
-            SourceKind::Git(_) => self.inner.canonical_url.hash(into),
-            _ => self.inner.url.as_str().hash(into),
-        }
+        self.inner.canonical_url.hash(into);
     }
```

Comment:
Most CanonicalUrl differences are related to git somehow.  I'm assuming that means we can't hit them in any other way.  There is the stripping of a trailing `/`.  However, we probably should have been respecting that in the hash in the first place I assume.  The question is if this will change anything significant enough

Reply:
Perhaps if you have a index hosted on git, then `GITHUB.com/me/my-index.git/` could change a lot. I think the biggest concern here is having to recheck out dependencies. So even if this does affect people we might just eat that cost. On the other hand there's a lot of complexity here and maybe there are bigger problems I forgot about.

Reply:
Haven't yet looked closer. I wonder if this would affect hacks like this one: <https://github.com/rust-lang/cargo/issues/5478#issuecomment-1055365283>.

Reply:
I don't think that would be affected by this change. But it is something we have to watch for as I "clean up" this code.

Reply:
That is true, since `[patch.<url>]` uses URLs directly, not SourceId.

This change may affect anywhere we let people define URL and parse it as `SourceId`, like the `[registries]` table? We could also deem this as a bug and fix it.

Reply:
I would hope that we rely on `Eq` for those cases and not `Hash`. So that the problem you identified is not a problem with this PR, rather one to investigate on the next PR that changes `Eq`.

Reply:
Fair enough. Have checked other usages of it, none of them seems user-facing like the path to index files. It should be fine to move over.